Add a doc outlining possible approach to SDK configuration.

anuraaga opened the pull request:
@jkwatson asked me to make a proposal of SDK configuration, and I've tried to outline goals and approaches here. This might not be the best approach, but it keeps these principles in mind

- Keep the end-user API as simple as possible
- It's ok to push some complexity downstream for less-common use cases like dynamic config, partial SDKs
- It's ok to expect more domain knowledge from authors of SDK components / extensions than of end users
- Keep core lightweight and let external configuration mechanisms take on the brunt of configuration

Perhaps these principles aren't good principles :) But this can hopefully give something to talk about.

anuraaga commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+ensured by the configuration framework such as Spring. For application developers, this restriction
+should not have any effect one way or the other in the vast majority of cases.
+
+MySQL is the only known corner case that requires the global SDK instance. If such a corner case
```
At least in Brave

bogdandrutu replied:
So there is no world outside Spring? :)

bogdandrutu commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+`Resource` for tracing and metrics, something that would be a gotcha. Exporters are handled natively
+given they are so important - even though the exporters are specified before span processors, they
+will always be executed last to make sure all span processors appy to them. The exporter builders
+inherit from a `ExporterBuilder` type of abstract class - it takes cafe of batching configuration to
```
s/cafe/care?

bogdandrutu commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+    .setTimeout(Duration.ofSeconds(10))
+    .setBatchQueueSize(10000)
+    .build())
+  .addSpanExporter(ZipkinExporter.builder()
+    .setEndpoint("spans-service:4317")
+    .disableBatching()
+    .build())
```
How can I set https://github.com/open-telemetry/opentelemetry-java/tree/master/sdk-extensions/async-processor to be used instead of the normal batch span processor?

anuraaga replied:
Oops forgot to reply - indeed this may make the fluent model too limiting. But ideally we can at least find a way to share timeout settings between processor and exporter, major pain right now.

jkwatson replied:
I think there's been some thought that maybe we should swap the responsibility and have the exporter be created with a processor (if the specific exporter allows for more than one option), rather than the other way around. That would definitely be a change to our processor/exporter APIs, but maybe now is a good time to think about it.

bogdandrutu commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+  }
+  
+  @Bean
+  public MeterProvider tracerProvider(OpenTelemetry openTelemetry) {
```
s/tracerProvider/meterProvider?

bogdandrutu commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+  private final Meter meter;
+  
+  @Inject
+  public MyService(TracerProvider tracerProvider, MeterProvider meterProvider) {
```
I would encourage passing Otel instance always. So if we add more functionality to the otel instance we don't break the API.

bogdandrutu commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+However, there are corner cases where an instance cannot be injected. The famous example is MySQL -
+MySQL interceptors are initialized by calling a default constructor, and there is no way to pass a
+built instance of a signal provider. For this case, we must store an SDK instance into a global
+variable. To reduce gotchas for application developers, the first SDK that is instantiated is stored
+automatically in the global variable.
```
I think this is a very unexpected behavior....

anuraaga replied:
It could also be reasonable to leave this up to frameworks - but it will almost always be required.

iNikem replied:
I agree with Bogdan. Let it be explicit.

jkwatson replied:
By "Let it be explicit" you mean "require the SDK user to explicitly assign the global"? Does this mean that the operator/app developer will need to understand the internals of all the instrumentation that happens to be included with the project?

I'm not necessarily arguing in favor of auto-setting the global instance, but I wanted to bring up a problem with the alternative.

My opinion would probably be: get rid of the global altogether and make instrumentation authors work around it in some way [I'm definitely not an expert here...just trying to optimize for simplicity as much as possible].

bogdandrutu commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+    components.add(sampler);
+  }
+
+  public OpenTelemetry build() {
+    OpenTelemetrySdk sdk = new OpenTelemetrySdk(tracerProvider.build(), meterProvider.build());
+    for (OpenTelemetryComponent component : components) {
+      component.setOpenTelemetry(sdk);
+    }
+  }
```
What if someone is very strict and want to actually use a different sdk instance + pipeline to monitor the main monitoring sdk?

bogdandrutu commented on docs/sdk-configuration.md:
````diff
@@ -0,0 +1,492 @@
+}
+```
+
+A framework author will have an even easier time since most dependency injection frameworks
+natively support lazy injection.
````
So if I am not using Spring I am not a "framework" author? :)

I would rewrite with something like:
When using a framework that supports dependency injection and lazy initialization ....

bogdandrutu commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+within telemetry components to those components, where the maintainers will have more domain
+knowledge. It allows this mutator method to be removed from the end-user API.
+
+- `updateTraceConfig` - instead of allowing updates at the top level, we should consider making
+`TraceConfig` an interface, and the SDK default implementation is static. It allows the above
+benefits of immutability to be in place for the common case where dynamic updates are not needed.
+Where dynamic updates are needed, it can be replaced with a mutable implementation instead of making
+the SDK configuration mutable. This keeps update methods out of the end-user APIs and will generally
+give framework developers more control by handling dynamicism themselves without the chance of
+end-users to affect it negatively.
```
If TraceConfig is not "immutable" you have hard time and undefined behaviors. What does it happen if we decide to call sampler twice and gets changed between the calls? Or any other property that we use?

bogdandrutu replied:
By making immutable we can guarantee that one config is applied for the entire lifetime of one span, and you cannot do this with a mutable config.

You may say that you add a "Factory" that returns an immutable TraceConfig, that is a lot of unnecessary allocations (you may cache the result). But if there is a factory why someone cannot do the same thing and update the Factory?

iNikem replied:
How are we going to support dynamic Samplers?

anuraaga replied:
Sampler is already an interface I was always expecting them to be a dynamic implementation of it.

iNikem replied:
All current Samplers are immutable and are changed by swapping instances.

jkwatson replied:
sure, but that doesn't mean that other Samplers could be more dynamic (or we could even update our existing Samplers to be dynamic without altering the interface, I think).

bogdandrutu commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+
+  public void doLogic() {
+    // My logic is important, so always sample it!
+    OpenTelemetrySdk.getTracerManagement().updateTraceConfig(config -> config.setSampler(ALWAYS_ON));
```
Who is the author of this line? Only one person should configure or control the SDK config. Any random "service" running in an application service should be able to affect only their instance of SDK if any, otherwise should not be able to control the config.

jkwatson replied:
It seems like this might be an argument against exposing a global where anyone could muck about with it and effect everything running in the current classloader (or JVM itself depending on where the global is class loaded first).

iNikem commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+- Uses the default Clock, which uses Java 8 / 9+ optimized APIs for getting time
+  - The only real reason a user would set this is for unit tests, not for production
+- Enforces default numeric limits related to number of attribute, etc
+
```
What about propagation?

anuraaga replied:
Thanks for reminder :)

iNikem commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+  .addSpanExporter(OtlpGrpcSpanExporter.builder()
+    .setEndpoint("spans-service:4317")
+    .setTimeout(Duration.ofSeconds(10))
+    .setBatchQueueSize(10000)
```
batching becomes the property of exporters?

anuraaga replied:
It's a random idea - more that batching configuration is the job of exporter configuration. One goal was to make it more obvious when an exporter is tied to a batching (or not) strategy.

But as with the comment about async processor this might be too limiting so not liking it so much anymore but still want to talk it through a bit.

iNikem commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+  .addSpanProcessor(CustomAttributeAddingProcessor.create())
+  .addSpanProcessor(CustomEventAddingProcessor.create())
+  .setTraceSampler(Sampler.rateLimiting())
+  .setTraceLimits(TraceLimits.builder().setMaxAttributes(10).build())
```
I like this

iNikem commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+Everything is exposed on the `OpenTelemetrySdkBuilder` to provide discoverability of all the parts
+of the SDK and to let the SDK manage shared components - it is not possible to configure a different
+`Resource` for tracing and metrics, something that would be a gotcha. Exporters are handled natively
+given they are so important - even though the exporters are specified before span processors, they
+will always be executed last to make sure all span processors appy to them. The exporter builders
```
How do you plan to chain `SpanProcessor.end()` calls?

anuraaga replied:
This proposal doesn't address the issue of not being able to have branching `SpanProcessor`. It just calls the processors in order, but trying to help with the exporter gotcha by making sure it comes at the end.

iNikem replied:
This seems to be as violating spec, et least formally

iNikem commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+OpenTelemetry to their app and see traces show up in a console. Application developers will increase
+in number forever, while the below are more constant.
+
+- Dev-ops / framework developers. Write components or frameworks to support providing tracing to
+their application developers. May write custom SDK extensions such as exporters, samplers, to fit
+with their internal infrastructure and as such have some familiarity with tracing at least as the
+SDK presents it.
+
```
I think we bundle two types here: authors of custom exporters/samplers and frameworks authors. The latter will handle configuration, but the former will just write a component according to an interface.

jkwatson replied:
the framework authors might need to do both (see Sleuth), but your point is still a good one.

iNikem commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+## The global instance of the SDK
+
+A built instance is convenient to use in most Java apps because of dependency injection. Because it
+has a easy-to-reason lifecycle, being tied into the dependency ordering (even if dependency 
```
I agree with this section, but not with "easy-to-reason lifecycle". Shared instance kept in a field will break if somebody somewhere calls `shutdown` on the same shared instance.

iNikem commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+MySQL is the only known corner case that requires the global SDK instance. If such a corner case
+didn't exist, we may not even support it in the first place.
+
+See special note with Java Agent below though.
```
```suggestion
See special note about Java Agent below though.
```

iNikem commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+  
+  @Override
+  public void setOpenTelemetry(OpenTelemetry openTelemetry) {
+    traer = openTelemetry.getTracerProvider().get("spanexporter");
```
```suggestion
    tracer = openTelemetry.getTracerProvider().get("spanexporter");
```

iNikem commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+
+## Auto-configuration
+
+The above presents programmatic configuration for the SDK and proposeds that the core SDK has no
```
```suggestion
The above presents programmatic configuration for the SDK and proposes that the core SDK has no
```

iNikem commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+## Auto-configuration
+
+The above presents programmatic configuration for the SDK and proposeds that the core SDK has no
+other mechanism for configuration. There is no SPI nor processing of environment variables or
```
Wait :) I understand your purpose and mostly support it, but configuration from environment is useful. If core SDK does not support (which is fine), we may need an example of how one can achieve it.

anuraaga replied:
Agree it is a decision on whether we prioritize the "wrapper artifact" I mentioned below. But do want to move it to a different layer.

iNikem commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+configuration wrapper, we avoid the possibility of confusion by having duplicate variables (in
+practice, OpenTelemetry naming would likely be ignored and overwritten by Spring naming).
+
+## Partial SDKs
```
I got very confused by this section. It seems like configuration duplication. What if I by mistake do both `OpenTelemetrySdk.builder().xxxxx` as above and `DefaultOpenTelemetry.builder().xxxx` as below?

anuraaga replied:
`OpenTelwetrySdk` is only present in the full SDK artifact (`-all`). So I guess it won't be there with a partial SDK?

Or is it potential confusion for full SDK users having both entry points? I don't think our repo should have any documentation related to `DefaultOpenTelemetry`. Maybe even hide it from JavaDoc. Full SDK users use `OpenTelemetrySdk` that's it, and it's all we support. Supporting partial SDK is up to the authors of partial SDK components IMO.

iNikem replied:
The second part, yes. So you are talking here about some other (partial) implementation of the SDK, not this one?

anuraaga replied:
Not sure what you're referring to with "this one" - my hypothetical micrometer SDK?

So I am imagining an app with dependencies on `opentelemetry-sdk-trace` and `opentelemetry-micrometer` and use `DefaultOpenTelemetry`.

Or there might be a `opentelemetry-sdk-with-micrometer` which depends on the above and provides `OpenTelemetrySdkWithMicrometerBuilder`. Obtuse but just an example of what I was thinking thinking here.

iNikem replied:
"this one" is "this repo". I mean, we provide full SDK and we don't plan to provide partial SDKs, right? Partial SDKs come from outside, right?

anuraaga replied:
Ah yeah that's what I'm thinking

jkwatson replied:
not true, though, since we'll be pulling metrics out of the "main" SDK and only providing that piece as an "experimental" signal until it's ready to go. We'll have a "partial" SDK for metrics, at least in some sense.

iNikem reviewed (commented):
I mostly support this proposal. Three open questions:

- Should java agent set Global SDK?
- Confusing part about partial SDKs
- Dynamic Samplers

jkwatson commented:
ref #1909 which I had forgotten existed.

jkwatson commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+- OpenTelemetry maintainers. Write the SDK code.
+
+When making decisions, especially about complexity, we always prioritize the application developers,
+then framework developers, then maintainers. This is because we expect those with less domain
```
this is an excellent call-out, and something we all definitely need to keep in mind. Sometimes this will mean we add "more than one way to do things" because it will make app developers lives easier, even if it makes our jobs as maintainers more annoying/painful.

jkwatson commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+
+- Reduce the chance of gotchas or configuration mishaps.
+
+- Aim for optimal performance for the most common use case.
```
is performance at configuration time really much of an issue and something we should worry about?

jkwatson commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+### Goals
+
+- Provide a single entrypoint to configuring the SDK. For end-users, less familiar with the SDK, we
+want to have everything together to provide discoverability and simpler end-user code.
```
is this a "single entrypoint" per use-case, or only a single entrypoint, period?

jkwatson commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+
+## Configuring an instance of the SDK
+
+The SDK consists of configuration options for all the signals it supports. Users all have different
```
```suggestion
The SDK exposes configuration options for all the signals it supports. Users all have different
```

jkwatson commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+## Configuring an instance of the SDK
+
+The SDK consists of configuration options for all the signals it supports. Users all have different
+requirements for how they use the SDK, for example they may use different exporters depending on
```
```suggestion
requirements for how they use the SDK; for example they may use different exporters depending on
```

jkwatson commented on docs/sdk-configuration.md:
````diff
@@ -0,0 +1,492 @@
+
+```java
+OpenTelemetrySdk openTelemetry = OpenTelemetrySdk.builder()
+  .addSpanExporter(OtlpGrpcSpanExporter.builder()
````
My opinion: no. But, I'm certainly up to being convinced otherwise. 

My reasoning is: people are going to try out the SDK on their local laptops first (probably), and if the default is to point at Otlp, they'll just get a bunch of errors in their logs.

I could be convinced that the default could be the logging exporter...that seems like an option that wouldn't break anybody and make it clear what's happening.

jkwatson commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+  .addSpanExporter(OtlpGrpcSpanExporter.builder()
+    .setEndpoint("spans-service:4317")
+    .build())
+  .addMetricExposition(PrometheusExposition.getDefault())
```
This example mixes two signals, which is good. The question is: are we likely to end up with an unwieldy builder if it contains the configuration for multiple signals, especially when we (eventually) add logs in here as well.

Would we be better off (as I think Bogdan has suggested) requiring the SDK components for each signal to be configured indivually, then have the `OpenTelemetrySdk` builder only compose the other components? I can see pros and cons.

jkwatson commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+      .setTraceLimits(TraceLimits.builder().setMaxAttributes(config.getMaxSpanAttributes()).build())
+      .setIdGenerator(TimestampedIdGenerator.create())
+      .setMeterStuff(InProgressMetricsStuff.create())
+      .build();
```
Here is a sticking point that I see with the omnibus build-everything-at-the-OpenTelemetrySdk-level: For 1.0, we need to remove metrics from the OpenTelemetrySdk, which will mean that we *must* provide a way to build the pre-release MeterSdkProvider in a stand-alone manner. And, there may (probably will) be people who only want to use one of the signals, without the omnibus OpenTelemetrySdk, so we'll need to provide individual builders for each of the signals.

If we need to provide individual builders for each SDK signal, then are we ok with having both ways of building things? If so, then 👍 . This does seem to be a case where we should favor the end-user and make sure that they can do it either way.

iNikem replied:
If we need to provide a way to configure and use just some of the signals, then I think that we should go this way:
> requiring the SDK components for each signal to be configured indivually, then have the OpenTelemetrySdk builder only compose the other components

The problem with supporting mix and match is this piece of code: https://github.com/open-telemetry/opentelemetry-java/pull/2231#issuecomment-741903911 .

jkwatson commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+  
+  @Override
+  public void export() {
+    tracer.spanBuilder("export").startSpan();
```
obviously would need a null check here for the tracer, but that's a nit, obviously

jkwatson commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,492 @@
+  return OpenTelemetrySdkWithMicrometer.builder()
+    .addSpanExporter()
+    .setMeterRegistry()
+    .build();
```
this is definitely something I hadn't pondered. This wouldn't work with pre-release signals, though, unless the pre-release metrics SDK (for instance) also provided an extension to the OpenTelemetry interface, too. 🤔

jkwatson commented:
1) Thanks for writing this up!!!
2) I think we also should consider about the configuration story with the "API only" users, where they only need to decide on propagators, and maybe some custom context backing store. Doesn't the question of global behavior of the SDK also hinge somewhat on the existence and use-cases for the global API in the absence of an SDK? In some sense, the API provides an SDK implementation, albeit a super minimal mostly no-op one. I feel like we should have the same story for configuring *that* minimal SDK that we do for a full-featured SDK.

jkwatson commented:
> 1. Thanks for writing this up!!!
> 2. I think we also should consider the configuration story with the "API only" users, where they only need to decide on propagators, and maybe some custom context backing store. Doesn't the question of global behavior of the SDK also hinge somewhat on the existence and use-cases for the global API in the absence of an SDK? In some sense, the API provides an SDK implementation, albeit a super minimal mostly no-op one. I feel like we should have the same story for configuring _that_ minimal SDK that we do for a full-featured SDK.

anuraaga commented:
I have rewritten the builder to accept signals instead of signal knobs - while the duplication of `Resource` pains me, I think it could be better than the alternative. I'm not 100% convinced though as that duplication did hurt to write it ;) I have fleshed out the auto-configuration wrapper, which I hope can satisfy the itch for the magic global for end users that like that. We would expect it not to be used alongside Spring, or the javaagent.

I have also removed some of the proposals around reducing the complexity of registering exporters - I think it's somewhat orthogonal to the "overall configuration story" and will require quite a bit of thought anyways.

anuraaga commented on docs/sdk-configuration.md:
````diff
@@ -0,0 +1,608 @@
+setting SDK implementations and is not meant for use with partial SDKs.
+
+```java
+class OpenTelemetrySdkBuilder {
````
Realized there's probably no point for this class.

jkwatson commented:
I think this is really solid. Thanks again for the work.  

I have only one thing I think we need to include and that is calling out how the "API only" case ends up working. We can't let it throw exceptions before the global has been set because library instrumentation might be using a global (against recommendations, but we can't stop it), and we can't let a library dependency crash an application. The user may not even know that the library they are using has been instrumented with OTel, and so we can't require the app dev to configure the API global.

jkwatson commented:
Oh...one more piece I'm not sure about...are we going to end up auto-setting the global the first time a full SDK has been built or no?

anuraaga commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,676 @@
+class OpenTelemetrySdkBuilder {
+  public OpenTelemetrySdkBuilder setTracerProvider(TracerSdkProvider tracerProvider);
+  public OpenTelemetrySdkBuilder setMeterProvider(MeterSdkProvider meterProvider);
+  public OpenTelemetrySdk buildAndSetAsGlobal();
```
@jkwatson I had removed the auto-set as global due to controversy, but do think it's the best option in most cases. If the problem is mystery and surprise, then this naming, while obtuse, can make it clear. I would not have a normal `build()` method as that would reduce the discoverability of this mechanism - it has to be `buildWithoutSetting` or something similar to make sure a user makes a conscious decision without just assuming to use `build()`.

jkwatson replied:
the only bit that might be confusing is that the 2nd one *wouldn't* set it as global, right?

anuraaga replied:
Yeah - I think you're saying we need to document these methods well?

jkwatson replied:
lol. yes, and maybe the 2nd call to `buildAndSetAsGlobal()` would fail? Not sure what the should be, but that can be decided at a later date.

anuraaga commented on docs/sdk-configuration.md:
````diff
@@ -0,0 +1,676 @@
+}
+```
+
+The `OpenTelemetrySdk` and `TracerSdkProvider` are still exposed to provide access to `forceFlush`.
````
@jkwatson I realized it could make sense to remove `shutdown` from the `TracerSdkProvider`. It seems like it could make some lifecycle woes simpler to reason about, maybe even enable `toBuilder()`?

anuraaga commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,676 @@
+Before an SDK has been set, access to the global `OpenTelemetry` will return a no-op 
+`DefaultOpenTelemetry`. This is because it is possible library instrumentation is using the global,
+and it may even use it during the processing of a request (rather than only at initialization time).
+For this reason, we cannot throw an exception. Instead, if the SDK is detected on the classpath, we
```
@jkwatson API-only case

jkwatson replied:
Thanks. This logging I think will be very helpful at finding where people are waiting until too late to set up their SDK. 👍

iNikem commented on docs/sdk-configuration.md:
````diff
@@ -0,0 +1,676 @@
+```java
+@Bean
+public OpenTelemetry openTelemetry() {
+  return DefaultOpenTelemetry.builder()
````
So what is the essential difference between this case and the "full" `OpenTelemetrySdkBuilder` above?

anuraaga replied:
The SDK builder returns types that expose forceFlush and shutdown as required by the spec. Hence me filing issues to try to get rid of those :-)

anuraaga commented on docs/sdk-configuration.md:
````diff
@@ -0,0 +1,676 @@
+}
+```
+
+## Immutability of OpenTelemetry configuration
````
Please let me know what's confusing since I want to clarify it :)

iNikem replied:
E.g. I don't understand the connection between `addSpanProcessor` method (on which class, btw?) and telemetry within telemetry (btw, is this the same as what you called "telemetry within SDK components" above?).

"the SDK default implementation is static" - do you mean "immutable"? And in general, sorry Anuraag, this section reads much sloppier than the preceding document :)

Also, I understand that Bogdan suggests having immutable SDK. As this does make total case on the first glance, may be this section better address those cases, when we potentially need mutability. E.g. "why obvious choice of immutable SDK is not that simple?"

anuraaga replied:
Thanks will try to rewrite it - one thing that's always tricky for me is talking about immutability. The SDK isn't immutable in the strictest sense since, e.g., `BatchSpanProcessor` has a mutable queue of spans. I guess the word to use is shallow-immutable which is what I meant by configuration being static. Let me see what I can do.

iNikem commented on docs/sdk-configuration.md:
````diff
@@ -0,0 +1,676 @@
+}
+```
+
+## TracerSdkManagement
````
And now I understand why :) This section does not talk about SDK configuration at all :) But some completely separate concerns of public interfaces of the SDK. Should this section be in document at all?

anuraaga replied:
This came to mind while considering what's left after removing the mutability methods from management interface, since ideally we don't have a management interface at all. But agree it's not really related with others, happy to have it just in commit history.

anuraaga commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,676 @@
+API as it's an end-user API - it could make the separation of concerns of the initialization of the
+SDK and the end-user tracing API complete. But the spec currently does not allow it.
+
+## Library instrumentation
```
Hmm - was I too nice to allow library authors an option of starting with global? I was imagining Bogdan's use case where some codebases may just want to make everything global, even in their instrumentation. But maybe it's too unclear in the context of this doc. Should I remove everything fro `An alternative method...`?

iNikem replied:
I think so, yes.

jkwatson commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,676 @@
+
+This configures resource, clock, exporters, span processors, propagators, sampler, trace limits, and metrics.
+It configures two SDKs with different propagators. We unfortunately cannot achieve our goal of only
+setting a property once - `Resource` and `Clock` are shared among signals but configured repeatedly.
```
@bogdandrutu has (I think) a use-case where sharing a `Resource` across MeterProviders will be prolematic. I'll let him explain.

iNikem commented:
Are we good to merge this? Or what is the plan with this?

jkwatson commented:
> Are we good to merge this? Or what is the plan with this?

I'll give this one last once-over before approving!

jkwatson commented on docs/sdk-configuration.md:
```diff
@@ -0,0 +1,612 @@
+  public OpenTelemetrySdkBuilder setTracerProvider(TracerSdkProvider tracerProvider);
+  public OpenTelemetrySdkBuilder setMeterProvider(MeterSdkProvider meterProvider);
+  public OpenTelemetrySdk buildAndSetAsGlobal();
+  public OpenTelemetrySdk buildWithoutSettingAsGlobal();
```
I might just call this one `build()`.  ;)

anuraaga replied:
See my note that I think this needs to be a conscious decision, though admittedly don't like the name much either :P

jkwatson reviewed (approved):
I think this is excellent. I suspect we'll evolve some bits of it as we go, but this seems like a fantastic starting point. Thanks again!
